### ISSUE TITLE
Memory leak:  Proxy.revocable's revoker holds the proxy strongly

### ISSUE THREAD

**Issue by ajvincent** [tc39/ecma262#2325]
https://262.ecma-international.org/11.0/#sec-proxy.revocable

Consider the following:
```javascript
const revokers = [];
{
  const {revoke} = Proxy.revocable({}, Reflect);
  revokers.push(revoke);
  // the created proxy is inaccessible, but the revoker holds it strongly in an internal slot
}
```
At first glance, this looks like a minor leak.  However, it might not be:
- Every proxy holds a strong reference to a shadow target
- Every proxy holds a strong reference to a proxy handler
- For a membrane, each object graph could hold hundreds or thousands of these revokers, as part of a design to revoke an entire object graph of proxies at once.  So the memory leak scales.  (Reminder:  membranes motivated adding proxies and weak maps to ECMA-262.)

**Comment by ajvincent** [comment]
@erights @caridy

**Comment by bakkot** [comment]
If an object is in fact not observable, the fact that the specification says to store it in some slot on some observable value does not prevent an engine from GC'ing it. As such I don't think it really makes sense to talk about internal slots as "strong" references.

**Comment by ljharb** [comment]
Agreed - I think that changing that internal slot to hold a WeakRef to the Proxy, and modifying the check in https://tc39.es/ecma262/#sec-proxy-revocation-functions to use https://tc39.es/ecma262/#sec-weakrefderef, would be an unobservable/non-normative change, which means that an engine is free to do precisely that with the current spec text.

**Comment by erights** [comment]
@ljharb thanks. We arrived at exactly the same conclusion at the SES meeting this morning.

**Comment by ajvincent** [comment]
It sounds like I misunderstood the specification - which would not be the first time!  At least this time it's a slight error, something deferred to the implementers of the specification.

I'm fine with that, at a basic level.  I'm just wondering if this issue should be closed as invalid, and/or what I might do to suggest implementers check for a possible leak (other than filing bugs that might be closed as "we already do that".)

I will admit, I'm not comfortable with the possibility of a membrane causing memory leaks even by accident.  I'm considering a countermeasure, `WeakMap<proxy, WeakRef<revoker>>` so that the proxy's existence is the only guarantee of the revoker's existence. 

For the case of multiple revokers based on one key (origin graph revoking proxies in other object graphs), I am considering `WeakMap<key, Set<WeakRef<revoker>>>`.  It's a bit ugly, but it should work.

I'm open to other ideas.  I just hope that I shouldn't need to worry about this.

**Comment by ajvincent** [comment]
Spidermonkey bug filed: https://bugzilla.mozilla.org/show_bug.cgi?id=1694781
V8 bug filed: https://bugs.chromium.org/p/v8/issues/detail?id=11507
